[Justin Siaw] ip

JustinS00 opened the pull request:
# Duke

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. install java
```java
public static void main(String[] args) {
	System.out.println("Hello World");
}
```
2. run Duke
3. type commands
4. see the magic 🪄 

## Features
- [X] Managing tasks
- [X] `delete` tasks
- [X] `find` tasks

Commit:
Add Gradle support

JustinS00 committed:
Add Greet, Echo, Exit

JustinS00 committed:
Add, List

JustinS00 committed:
Mark as Done

JustinS00 committed:
Add Todo, Event, Deadline

JustinS00 committed:
Add TextUiTesting

JustinS00 committed:
Handle Errors

JustinS00 committed:
Add Delete Task

JustinS00 committed:
Save System

JustinS00 committed:
Add DateTime for Deadline

JustinS00 committed:
Merge branch 'branch-Level-8'

JustinS00 committed:
Add more testcases for text-ui-test

JustinS00 committed:
Add Enums for Command and more OOP

JustinS00 committed:
Refactor into packages

JustinS00 committed:
Rename SaveManager to Storage

JustinS00 committed:
Refactor Duke Ui

JustinS00 committed:
runtest.bat: delete previous run data

JustinS00 committed:
Add Parser class

JustinS00 committed:
Updated dukeShowList() to justify numbers dynamically

JustinS00 committed:
Merge branch 'add-gradle-support'

JustinS00 committed:
Add Gradle for build and run

JustinS00 committed:
Add JUnit

JustinS00 committed:
Add Java Docs

JustinS00 committed:
Add Find command

JustinS00 committed:
Update variable names

JustinS00 committed:
Merge branch 'branch-Level-9'

JustinS00 committed:
Fix merge conflict

JustinS00 committed:
Fix minor bugs in output

JustinS00 committed:
Fix find bug, Add more JavaDoc

JustinS00 committed:
Fix more styling issues

comicalromance commented on src/main/java/duke/command/DukeCommandType.java:
```diff
@@ -0,0 +1,17 @@
+/**
+ * Enumeration of the type of commands available in Duke
+ */
+public enum DukeCommandType {
```
Great use of enum in a separate class to clarify what each method is about!

albertZhangTJ replied:
I really like how you used enum to differentiate all kinds of commands.

comicalromance commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,55 @@
+        try {
+            FileInputStream fin = new FileInputStream(FILE_NAME);
+            ObjectInputStream ois = new ObjectInputStream(fin);
+            @SuppressWarnings("unchecked") ArrayList<Task> tasks = (ArrayList<Task>) ois.readObject();
```
It'd be great if you could also close the FileInputStream after you're done with it (a 'try with resources' statement would be very convenient for your usecase!)

comicalromance commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,58 @@
+        case "find":
+            return DukeCommandType.FIND;
+        default:
+            return DukeCommandType.UNKNOWN;
```
I really like how you handled the default case instead of just leaving it to fail / cause an error!

comicalromance commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,81 @@
+            DukeCommandType commandType = command.getCommandType();
+            String args = command.getArgs();
+            switch (commandType) {
+            case EXIT: {
```
Might be personal preference but the braces on switch statements are optional and so you could technically remove them if you want to! Same goes for the other switch statements as well.

comicalromance commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,59 @@
+
+    private String parseDateTime(String originalString) {
+        StringBuffer deadline = new StringBuffer(originalString.replace("/", "-"));
+        if (Pattern.compile("([0-9]{4})[-]([1][0-2]|0[1-9])[-](3[01]|[12][0-9]|0[1-9])\\s[0-2][0-9][0-6][0-9]")
+                .matcher(deadline).matches()) { //yyyy-mm-dd HHmm
+            deadline.insert(deadline.length() - 2, ":");
+            deadline.replace(deadline.indexOf(" "), deadline.indexOf(" ") + 1, "T");
+            return LocalDateTime.parse(deadline).format(DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm"));
+        } else if (Pattern.compile("(3[01]|[12][0-9]|0[1-9])[-]([1][0-2]|0[1-9])[-][0-9]{4}\\s[0-2][0-9][0-6][0-9]")
+                .matcher(deadline).matches()) { //dd-mm-yyyy HHmm
+            deadline.insert(deadline.length() - 2, ":");
+            deadline.replace(deadline.indexOf(" "), deadline.indexOf(" ") + 1, "T");
+            String dd = deadline.substring(0, 2);
+            deadline.replace(0, 2, deadline.substring(6, 10)).replace(8, 12, dd);
+            return LocalDateTime.parse(deadline).format(DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm"));
+        } else if (Pattern.compile("([0-9]{4})[-]([1][0-2]|0[1-9])[-](3[01]|[12][0-9]|0[1-9])")
+                .matcher(deadline).matches()) { //yyyy-mm-dd
+            return LocalDate.parse(deadline).format(DateTimeFormatter.ofPattern("dd MMM yyyy"));
+        } else if (Pattern.compile("(3[01]|[12][0-9]|0[1-9])[-]([1][0-2]|0[1-9])[-][0-9]{4}")
+                .matcher(deadline).matches()) { //dd-mm-yyyy
+            String dd = deadline.substring(0, 2);
+            deadline.replace(0, 2, deadline.substring(6, 10)).replace(8, 12, dd);
+            return LocalDate.parse(deadline).format(DateTimeFormatter.ofPattern("dd MMM yyyy"));
+        }
```
This section may be too complicated (especially the if statements). Would it help if you perhaps initialize the patterns as individual variables first (with names to indicate the pattern used) and then use it in the conditions?

comicalromance reviewed (commented):
LGTM! The structure and flow of the code seems very functional and I feel the OOP has been definitely executed well.

JustinS00 committed:
Add basic GUI

JustinS00 committed:
Refine GUI

albertZhangTJ reviewed (commented):
I really like your work! I think its quite sophisticatedly coded and it is really hard for me to find style errors!
 One suggestion I can come up with is you might want to add `@author` tags to your class-level JavaDoc comments

JustinS00 committed:
Add Assertions

There are no assertions to verify the assumptions made in the code.

Assertion failures indicates a possible bug in the code and will be
useful in debugging.

To ensure the code base is running as intended, let's add assertions
statements where appropriate.

JustinS00 committed:
Merge pull request #1 from JustinS00/branch-A-Assertions

Add Assertions

JustinS00 committed:
Refactor code to follow coding standards

Some parts of the code is not following the coding standards.

To ensure consistency and better readibility.

Lets refactor the code to adhere to the coding standard.

JustinS00 committed:
Merge pull request #2 from JustinS00/branch-A-CodeQuality

Refactor code to follow coding standards

JustinS00 committed:
Add gradle.yml

JustinS00 committed:
gradle.yml: remove i/o redirection

JustinS00 committed:
Add update command

JustinS00 committed:
Fix styling issues

JustinS00 committed:
Update User Guide

JustinS00 committed:
docs/README.md - Add Table Of Content and Quick Start

JustinS00 committed:
Update README.md

JustinS00 committed:
Update error message
